File: app/src/main/java/com/stevenfrew/beatprompter/Task.java
package com.stevenfrew.beatprompter;

import android.util.Log;

public abstract class Task implements Runnable
{
    private boolean mRunning;
    private boolean mStop=false;
    private final Object runningSync=new Object();
    private final Object stopSync=new Object();

    private final static String TASKTAG="task";

    public Task(boolean initialRunningState)
    {
        mRunning=initialRunningState;
    }
    private boolean getIsRunning()
    {
        synchronized (runningSync)
        {
            return mRunning;
        }
    }
    private void setIsRunning(boolean value)
    {
        synchronized (runningSync)
        {
            mRunning=value;
        }
    }
    protected boolean getShouldStop()
    {
        synchronized (stopSync)
        {
            return mStop;
        }
    }
    protected void setShouldStop(boolean value)
    {
        synchronized (stopSync)
        {
            mStop=value;
        }
    }
    public void run()
    {
        Log.d(TASKTAG,"Task initialising.");
        initialise();
        Log.d(TASKTAG,"Task starting.");
        while(!getShouldStop())
        {
            if(getIsRunning())
            {
                doWork();
            }
            else
            {
                try
                {
                    Thread.sleep(500);
                }
                catch(InterruptedException ie)
                {
                    Log.d(TASKTAG,"MIDI thread sleep (while paused) was interrupted.",ie);
                }
            }
        }
        Log.d(TASKTAG,"Task ending ... cleaning up.");
        cleanup();
        Log.d(TASKTAG,"Task ended.");
    }
    void stop()
    {
        setIsRunning(false);
        setShouldStop(true);
    }
    void initialise()
    {
    }
    void cleanup()
    {
    }
    private void pause()
    {
        setIsRunning(false);
    }
    private void resume()
    {
        setIsRunning(true);
    }
    public abstract void doWork();

    static void pauseTask(Task task,Thread thread) {
        if(task!=null) {
            task.pause();
            if(thread!=null)
                thread.interrupt();
        }
    }

    static void resumeTask(Task task) {
        if(task!=null) {
            task.resume();
        }
    }

    static void stopTask(Task task,Thread thread) {
        if(task!=null) {
            task.stop();
            if(thread!=null) {
                thread.interrupt();
                try {
                    thread.join();
                } catch (InterruptedException ie) {
                    Log.d(TASKTAG, "Task interrupted while waiting for join.", ie);
                }
            }
        }
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/Task.kt
package com.stevenfrew.beatprompter

import android.util.Log

abstract class Task(private var running: Boolean) : Runnable {
	private var stop = false
	private val runningSync = Any()
	private val stopSync = Any()
	private var isRunning: Boolean
		get() = synchronized(runningSync) {
			return running
		}
		set(value) = synchronized(runningSync) {
			running = value
		}
	protected val shouldStop: Boolean
		get() = synchronized(stopSync) {
			return stop
		}

	private fun setShouldStop() =
		synchronized(stopSync) {
			stop = true
		}

	override fun run() {
		Log.d(TASK_TAG, "Task initialising.")
		initialise()
		Log.d(TASK_TAG, "Task starting.")
		while (!shouldStop) {
			if (isRunning) {
				doWork()
			} else {
				try {
					Thread.sleep(500)
				} catch (ie: InterruptedException) {
					Log.d(TASK_TAG, "Thread sleep (while paused) was interrupted.", ie)
				}
			}
		}
		Log.d(TASK_TAG, "Task ended.")
	}

	private fun pause(): Boolean {
		val wasRunning = isRunning
		isRunning = false
		return wasRunning
	}

	private fun resume(): Boolean {
		val wasRunning = isRunning
		isRunning = true
		return !wasRunning
	}

	open fun stop(): Boolean {
		val wasRunningAndNotStopping = isRunning && !shouldStop
		isRunning = false
		setShouldStop()
		return wasRunningAndNotStopping
	}

	protected open fun initialise() {}
	abstract fun doWork()

	companion object {
		private const val TASK_TAG = "task"

		private fun changeTaskState(task: Task?, thread: Thread?, fn: (Task) -> Boolean): Boolean =
			if (task != null) {
				if (fn(task)) {
					if (thread != null) {
						thread.interrupt()
						true
					} else false
				} else false
			} else false

		fun pauseTask(task: Task?, thread: Thread?) = changeTaskState(task, thread) { it.pause() }
		fun resumeTask(task: Task?, thread: Thread?) = changeTaskState(task, thread) { it.resume() }

		fun stopTask(task: Task?, thread: Thread?) {
			if (changeTaskState(task, thread) { it.stop() })
				try {
					thread?.join()
				} catch (ie: InterruptedException) {
					Log.d(TASK_TAG, "Task interrupted while waiting for join.", ie)
				}
		}
	}
}